### PR TITLE
Various tweaks backported from permissions :lock:

### PR DESCRIPTION
**Pull request by camsaul**
As usual while working on unrelated things I have found numerous opportunities to improve and clean up various sections of code. The permissions branch is currently hovering around 3000 LOC so I am breaking out these unrelated improvements into a new branch

The highlights:
-  Use `console.error` a few places to log errors instead of `console.log`
-  `if (last_name !== 'undefined')` is missing a `typeof` (etc.) 😉 
-  Eliminate separate DB call to fetch superuser status in a significant chunk of our API calls. This is now fetched as part of the same call that fetches user ID associated with Metabase session. Across-the-board performance improvement!
-  Name things using `lisp-case` in the backend, not `snake_case`
-  Fix more functions that should have ended in an exclamation point from day one
-  Don't forget to `require` a namespace before using it
-  Fix some docstrings

### CODE REVIEW

File: frontend/src/metabase/components/UserAvatar.jsx [hunk 1]
```diff
@@ -26,17 +26,13 @@ export default class UserAvatar extends Component {
...
+            return typeof name !== 'undefined' && name.length ? name.substring(0, 1).toUpperCase() : '';
         }
 
-        if (last_name !== 'undefined') {
-            initials = initials + last_name.substring(0, 1).toUpperCase();
-        }
+        const initials = initial(first_name) + initial(last_name);
 
-        return initials;
+        return initials.length ? initials : '?';
```
Comment (salsakran) [inline frontend/src/metabase/components/UserAvatar.jsx]: can we add a test for this? ;)

File: src/metabase/api/card.clj [hunk 2]
```diff
@@ -78,18 +77,18 @@
   (cards-with-ids (map :model_id (db/select [ViewLog :model_id [:%max.timestamp :max]]
                                    :model   "card"
                                    :user_id *current-user-id*
-                                   (-> (h/group :model_id)
-                                       (h/order-by [:max :desc])
-                                       (h/limit 10))))))
+                                   {:group-by [:model_id]
+                                    :order-by [[:max :desc]]
+                                    :limit    10}))))
```
Comment (salsakran) [inline src/metabase/api/card.clj]: much easier on the eyes.

File: src/metabase/api/common.clj [hunk 3]
```diff
@@ -26,6 +26,10 @@
    ex. `@*current-user*`"
   (atom nil)) ; default binding is just something that will return nil when dereferenced
 
+(def ^:dynamic ^Boolean *is-superuser?*
+  "Is the current user a superuser?"
+  false)
+
```
Comment (salsakran) [inline src/metabase/api/common.clj]: I'm assuming this gets overloaded somewhere?

File: src/metabase/middleware.clj [hunk 4]
```diff
@@ -60,14 +60,18 @@
...
-                     (let [session-age-ms (- (System/currentTimeMillis) (.getTime ^java.util.Date (get session :created_at (u/->Date 0))))]
+                     (let [session-age-ms (- (System/currentTimeMillis) (or (when-let [^java.util.Date created-at (:created_at session)]
+                                                                              (.getTime created-at))
+                                                                            0))]
                        ;; If the session exists and is not expired (max-session-age > session-age) then validation is good
                        (when (and session (> (config/config-int :max-session-age) (quot session-age-ms 60000)))
-                         (assoc request :metabase-user-id (:user_id session))))))
+                         (assoc request
+                           :metabase-user-id (:user_id session)
+                           :is-superuser?    (:is_superuser session))))))
```
Comment (salsakran) [inline src/metabase/middleware.clj]: Does this mean that if a user has a session running and their account is stripped of the superuser bit, we'd need to wait `max-session-age` minutes until they can't fuck things up?

File: src/metabase/api/common.clj (hunk 3 above)
Comment (camsaul) [inline src/metabase/api/common.clj]: Yup the API call middleware binds this to `true`/`false` as appropriate for the current user

**Comment by salsakran** [comment]
I think if someone loses the admin bit, we should either nuke their existing sessions, or alternatively update their superuser bit in the session. 

Agree that this is a very low frequency event and we shouldn't be doing a DB round trip on every request for it.

File: src/metabase/middleware.clj (hunk 4 above)
Comment (camsaul) [inline src/metabase/middleware.clj]: No this is fetched for every API call

**Comment by camsaul** [comment]
We do a DB round-trip on every API call to make sure the credentials they used are valid and belong to an active user. This required a join to the user table.

We used to do a second-round trip on some API calls to check whether the current user had superuser status. The only change I made is including `is_superuser` in the join done every time so we don't need to do the second round-trip

**Comment by camsaul** [comment]
`is_superuser` isn't part of Session, it's part of User

**Comment by salsakran** [comment]
gotcha!

File: src/metabase/api/common.clj (hunk 3 above)
Comment (camsaul) [inline src/metabase/api/common.clj]: @salsakran variables with earmuffs around them are dynamic variables, which are a magical Lisp feature you don't really see in other languages.

```clojure
(def ^:dynamic *max-rows* 100)

(defn x []
  (println *max-rows*)) 

(x) ; -> 100

(binding [*max-rows* 200]
  (x)) ; -> 200

(x) ; -> 100
```

In Clojure they're thread-local so millions of API calls can be going on at the same time and everyone will have the appropriate value of `*is-superuser?*` when the look at it

**Comment by camsaul** [comment]
👍  ?

**Comment by salsakran** [comment]
sure you can have a thumb 👍

**Pull request merged**